cloud-storage-reference-app.md

```diff
@@ -0,0 +1,366 @@
+    CHCS-IAM Web App->>Patient Peter CH App: Access Token<br/>& Refresh Token
+    Note right of Patient Peter CH App: Token contain user.id<br/>and consented scopes
+    Patient Peter CH App->>Patient Peter CH App: Vendor Sign-in and Sharing Consent
+    Patient Peter CH App->>Patient Peter CH App: Periodic Vendor Data Download
+    Patient Peter CH App->>CHCS-IAM Web App: Uploads Data with<br/>OAuth2 Access Token
+    Researcher Rachel->>CHCS-IAM Web App: Views/sorts/downloads<br/>Patient Peter data
+```
+
+## SMART Client Flow
+
+```mermaid
+sequenceDiagram
+    actor Dr Debora
+    participant EHR
+    participant SMART CHCS
+    Dr Debora->>EHR: Seaches Patient Peter<br/>Clicks to launch CH
+    EHR->>SMART CHCS: GET chcs.org/launch<br/>?iss=https://fhir.example.com&launch=xyz123
+    SMART CHCS->>EHR: GET fhir.ehr.com/.well-known/smart-configuration
+    EHR->>SMART CHCS: Conformance JSON<br/>grant_types_supported, authorization_endpoint,<br/>jwks_uri, token_endpoint, capabilities,<br/>code_challenge_methods_supported
+    SMART CHCS->>EHR: GET auth.ehr.com<br/>?client_id=abc123&launch=xyz123<br/>&state=999&scope=launch+patient<br/>&redirect_uri=https://chcs.org/redirect&...
+    EHR->>SMART CHCS: GET chcs.org/redirect<br/>?code=1111&state=999
+    SMART CHCS->>EHR: POST<br/>code=1111, grant_type=access_token,scope=launch+patient,<br/>code_verifier=0f01, redirect_uri=https://chcs.org/redirect
+    EHR->>SMART CHCS: Auth JSON<br/>access_token, token_type, scope, patient=def123.456
+    SMART CHCS->>EHR: GET /fhir/patients/def.456<br/>Bearer: access_token
+    SMART CHCS->>SMART CHCS: Matches EHR user.id<br/>to CHCS user
+    EHR->>SMART CHCS: First Name, Last Name, Cell, E-mail
+    SMART CHCS->>SMART CHCS: Matches EHR patient.id<br/>to CHCS patient
+    SMART CHCS->>SMART CHCS: (Continues flow with Patient Peter<br/>SMART CHCS as above)
+    SMART CHCS->>Dr Debora: Dr Debora signed in<br/>Patient Peter shared personal device data displayed
+```
+
+______________________________________________________________________
+
+# Reference Material
+
+## Considerations
+
+### Cloud Storage Service(CSS) as a SMART Server
+
+- SMART Cloud Storage Service is source of truth for auth (with exception of SAML/SSO)
+- For access control against personal device data Cloud Storage Service should also be store of records, otherwise will require double-auth or some kind of pre-signed URIs and additional complexity unnecessary for a reference implementation
+- Without Secondary App access, auth is really just OAuth2 and not SMART
+
+```mermaid
+sequenceDiagram
+    actor Dr Debora
+    participant SMART CSS
+    participant Provider Directory Service
+    participant Secondary App (SMART Client)
+    Dr Debora->>SMART CSS: css.org<br/>Sign up / Sign in
+    opt
+        SMART CSS->>Provider Directory Service: SAML/SSO Auth
+        Provider Directory Service->>SMART CSS:
```

Comment:
MyST's mermaid support needs a return message description following the `:` here, e.g. `Provider Directory Service->>SMART CSS: Auth response`.

Reply:
Same issue as above

Reply:
Thanks @ryanlovett - ahhhhh I see so there is supposed to be a white space char following the ":" that must have been trimmed along the way. I just checked and that works with MyST ✅ 

@surfdoc if you replace everything between the `` ```mermaid `` with the code below it should work (and will fix the related below issue too)

```
sequenceDiagram
    actor Dr Debora
    participant SMART CSS
    participant Provider Directory Service
    participant Secondary App (SMART Client)
    Dr Debora->>SMART CSS: css.org<br/>Sign up / Sign in
    opt
        SMART CSS->>Provider Directory Service: SAML/SSO Auth
        Provider Directory Service->>SMART CSS: 
    end
    Dr Debora->>SMART CSS: Set up study, users and groups
    Dr Debora->>SMART CSS: Register Patient, create and send deep link
    Dr Debora->>Secondary App (SMART Client): Click to launch
    Secondary App (SMART Client)->>SMART CSS: SMART Auth flow
    SMART CSS->>Secondary App (SMART Client):  
    Secondary App (SMART Client)->>SMART CSS: GET /fhir/metadata
    SMART CSS->>Secondary App (SMART Client): Describe Observation valueAttachments
    Secondary App (SMART Client)->>SMART CSS: GET /observation?patient=123456
    SMART CSS->>Secondary App (SMART Client): Observation valueAttachments
    Secondary App (SMART Client)->>Dr Debora: Display shared personal device data
```

Reply:
@ryanlovett - There is a pre-commit that runs and removes trailing whitespaces. I think it is trimming these automatically.

Reply:
@minrk What do you think about disabling the pre-commit check for trailing whitespace?

Reply:
Sure, no problem!

Reply:
Or alternatively we can just repeat the label as per Ryan's original suggestion if that's easier? I wouldn't want you guys changing your whole pipeline just for this one diagram!

Reply:
Would coming up with new labels or repeating the previous ones for the returns be straightforward?

If this is a needed pattern for this and future diagrams then it isn't too hard to disable the pre-commit check.

Reply:
@s1monj and @ryanlovett  - I have been summoned for jury duty this week so will have limited availability, however, let me know what you decide and I will make the appropriate changes and update the PR. Thanks.

Reply:
@ryanlovett yep for sure - I'll just get rid of the second line it's not needed. Also looks like mermaid has just added double-headed arrows so will prob be available in Github in a few weeks.

@surfdoc if you can please replace with code below thanks
```
sequenceDiagram
    actor Dr Debora
    participant SMART CSS
    participant Provider Directory Service
    participant Secondary App (SMART Client)
    Dr Debora->>SMART CSS: css.org<br/>Sign up / Sign in
    opt
        SMART CSS->>Provider Directory Service: SAML/SSO Auth
        Provider Directory Service->>SMART CSS: 
    end
    Dr Debora->>SMART CSS: Set up study, users and groups
    Dr Debora->>SMART CSS: Register Patient, create and send deep link
    Dr Debora->>Secondary App (SMART Client): Click to launch
    Secondary App (SMART Client)->>SMART CSS: SMART Auth flow
    SMART CSS->>Secondary App (SMART Client):  
    Secondary App (SMART Client)->>SMART CSS: GET /fhir/metadata
    SMART CSS->>Secondary App (SMART Client): Describe Observation valueAttachments
    Secondary App (SMART Client)->>SMART CSS: GET /observation?patient=123456
    SMART CSS->>Secondary App (SMART Client): Observation valueAttachments
    Secondary App (SMART Client)->>Dr Debora: Display shared personal device data
```

Reply:
@s1monj I updated the mermaid diagrams removing lines where there was only a white space after the colon for the return arrow. I also made the requested changes if you want to re-review the PR. Thanks.

Reply:
@surfdoc Thanks, this renders correctly in MyST now! @s1monj if it looks good I think we can merge.